### ISSUE TITLE
Question on ImplicitMotionAlignment

### ISSUE THREAD

**Issue by johndpope** [hologerry/IMF#3]
im stuck on the hardest part. 
Been pulling my hair out. 

are we flattening p_k / p_q so it's all 1d vector? 
is by concatenating?

if i attempt to use the feature / motion dimesions - 
they're huge. the self attention blows up. 

        key_dim = ml_r.size(1) # 1048576
        value_dim = fl_r.size(1) # 1048576

https://github.com/johndpope/IMF

```shell
⚾ DenseFeatureEncoder input shape: torch.Size([1, 3, 256, 256])
    After initial conv: torch.Size([1, 64, 256, 256])
    After down_block 1: torch.Size([1, 64, 128, 128])
    After down_block 2: torch.Size([1, 128, 64, 64])
    After down_block 3: torch.Size([1, 256, 32, 32])
    After down_block 4: torch.Size([1, 512, 16, 16])
    After down_block 5: torch.Size([1, 512, 8, 8])
    DenseFeatureEncoder output shapes: [torch.Size([1, 64, 128, 128]), torch.Size([1, 128, 64, 64]), torch.Size([1, 256, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]
💳 LatentTokenEncoder input shape: torch.Size([1, 3, 256, 256])
    After first conv, bn, relu: torch.Size([1, 64, 128, 128])
    After resblock1: torch.Size([1, 64, 128, 128])
    After resblock2: torch.Size([1, 128, 64, 64])
    After resblock3: torch.Size([1, 256, 32, 32])
    After resblock4: torch.Size([1, 512, 16, 16])
    After equal_conv: torch.Size([1, 512, 16, 16])
    After adaptive_pool: torch.Size([1, 512, 1, 1])
    After flatten: torch.Size([1, 512])
    1xdm=32 LatentTokenEncoder output shape: torch.Size([1, 32])
💳 LatentTokenEncoder input shape: torch.Size([1, 3, 256, 256])
    After first conv, bn, relu: torch.Size([1, 64, 128, 128])
    After resblock1: torch.Size([1, 64, 128, 128])
    After resblock2: torch.Size([1, 128, 64, 64])
    After resblock3: torch.Size([1, 256, 32, 32])
    After resblock4: torch.Size([1, 512, 16, 16])
    After equal_conv: torch.Size([1, 512, 16, 16])
    After adaptive_pool: torch.Size([1, 512, 1, 1])
    After flatten: torch.Size([1, 512])
    1xdm=32 LatentTokenEncoder output shape: torch.Size([1, 32])
🎑 LatentTokenDecoder input shape: torch.Size([1, 32])
    After const: torch.Size([1, 64, 4, 4])
    After style_conv 1: torch.Size([1, 512, 4, 4])
    After style_conv 2: torch.Size([1, 512, 8, 8])
    After style_conv 3: torch.Size([1, 512, 8, 8])
    After style_conv 4: torch.Size([1, 512, 8, 8])
    After style_conv 5: torch.Size([1, 512, 16, 16])
    After style_conv 6: torch.Size([1, 512, 16, 16])
    After style_conv 7: torch.Size([1, 512, 16, 16])
    After style_conv 8: torch.Size([1, 512, 32, 32])
    After style_conv 9: torch.Size([1, 512, 32, 32])
    After style_conv 10: torch.Size([1, 512, 32, 32])
    After style_conv 11: torch.Size([1, 256, 64, 64])
    After style_conv 12: torch.Size([1, 256, 64, 64])
    After style_conv 13: torch.Size([1, 256, 64, 64])
    LatentTokenDecoder output shapes: [torch.Size([1, 256, 64, 64]), torch.Size([1, 512, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]
🎑 LatentTokenDecoder input shape: torch.Size([1, 32])
    After const: torch.Size([1, 64, 4, 4])
    After style_conv 1: torch.Size([1, 512, 4, 4])
    After style_conv 2: torch.Size([1, 512, 8, 8])
    After style_conv 3: torch.Size([1, 512, 8, 8])
    After style_conv 4: torch.Size([1, 512, 8, 8])
    After style_conv 5: torch.Size([1, 512, 16, 16])
    After style_conv 6: torch.Size([1, 512, 16, 16])
    After style_conv 7: torch.Size([1, 512, 16, 16])
    After style_conv 8: torch.Size([1, 512, 32, 32])
    After style_conv 9: torch.Size([1, 512, 32, 32])
    After style_conv 10: torch.Size([1, 512, 32, 32])
    After style_conv 11: torch.Size([1, 256, 64, 64])
    After style_conv 12: torch.Size([1, 256, 64, 64])
    After style_conv 13: torch.Size([1, 256, 64, 64])
    LatentTokenDecoder output shapes: [torch.Size([1, 256, 64, 64]), torch.Size([1, 512, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]
🎣 IMF - Number of layers: 4
f_r shapes: [torch.Size([1, 64, 128, 128]), torch.Size([1, 128, 64, 64]), torch.Size([1, 256, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]
m_r shapes: [torch.Size([1, 256, 64, 64]), torch.Size([1, 512, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]
m_c shapes: [torch.Size([1, 256, 64, 64]), torch.Size([1, 512, 32, 32]), torch.Size([1, 512, 16, 16]), torch.Size([1, 512, 8, 8])]

Processing layer 1:
  f_r_i shape: torch.Size([1, 64, 128, 128])
  m_r_i shape: torch.Size([1, 256, 64, 64])
  m_c_i shape: torch.Size([1, 256, 64, 64])
🎮 ImplicitMotionAlignment input shapes:
  ml_c: torch.Size([1, 256, 64, 64])
  ml_r: torch.Size([1, 256, 64, 64])
  fl_r: torch.Size([1, 64, 128, 128])
Before flattening:
  ml_c: torch.Size([1, 256, 64, 64])
  ml_r: torch.Size([1, 256, 64, 64])
  fl_r: torch.Size([1, 64, 128, 128])
  p_q: torch.Size([1, 64, 64, 64])
  p_k: torch.Size([1, 64, 64, 64])
After flattening and adding positional embeddings:
  ml_c: torch.Size([1, 1310720])
  ml_r: torch.Size([1, 1310720])
  fl_r: torch.Size([1, 1048576])
  p_q: torch.Size([1, 262144])
  p_k: torch.Size([1, 262144])
```

```python
class ImplicitMotionAlignment(nn.Module):
    def __init__(self, motion_dim, feature_dim, num_heads=8):
        super(ImplicitMotionAlignment, self).__init__()
        self.num_heads = num_heads
        
        self.pq = PositionalEncoding1D(motion_dim) # not feature_dim! motion features ml_c and ml_r as the queries (Q) and keys (K)
        self.pk = PositionalEncoding1D(motion_dim)
        
        self.cross_attention = nn.MultiheadAttention(embed_dim=motion_dim, num_heads=num_heads, batch_first=True)
        
        self.transformer_blocks = nn.ModuleList([
            TransformerBlock(motion_dim, num_heads) for _ in range(4)
        ])
        
        self.output_proj = nn.Linear(motion_dim, feature_dim)
        self.input_proj = nn.Linear(feature_dim, motion_dim)

    def forward(self, ml_c, ml_r, fl_r):
        b, c, h, w = ml_c.shape # for reprojection
        print("🎮 ImplicitMotionAlignment input shapes:")
        print(f"  ml_c: {ml_c.shape}")
        print(f"  ml_r: {ml_r.shape}")
        print(f"  fl_r: {fl_r.shape}")

        # Get the device of the input tensor
        device = ml_c.device

   

        print("Before flattening:")
        print(f"  ml_c: {ml_c.shape}")
        print(f"  ml_r: {ml_r.shape}")
        print(f"  fl_r: {fl_r.shape}")


        # create positional embeddings
        p_q = self.pq(ml_c)
        p_k = self.pk(ml_r)
        print(f"  p_q: {p_q.shape}")
        print(f"  p_k: {p_k.shape}")


        # Flatten into a 1D
        ml_c = ml_c.flatten(1, 3)
        ml_r = ml_r.flatten(1, 3)
        fl_r = fl_r.flatten(1, 3)

        # 🤷 flatten this too ? 
        p_q = p_q.flatten(1, 3)
        p_k = p_k.flatten(1, 3)
        # Add positional embeddings
        # ml_c = ml_c + p_q - this blows up
        # ml_r = ml_r + p_k
        # 🤷 Concatenate instead of add???
        ml_c = torch.cat([ml_c, p_q], dim=-1)
        ml_r = torch.cat([ml_r, p_k], dim=-1)


        print("After flattening and adding positional embeddings:")
        print(f"  ml_c: {ml_c.shape}")
        print(f"  ml_r: {ml_r.shape}")
        print(f"  fl_r: {fl_r.shape}")
        print(f"  p_q: {p_q.shape}")
        print(f"  p_k: {p_k.shape}")
        # Create CrossAttention on the fly with correct dimensions
        query_dim = ml_c.size(1) # 🤷 this is right?  1048576
        key_dim = ml_r.size(1) # 1048576
        value_dim = fl_r.size(1) # 1048576


        cross_attention = CrossAttention(query_dim, key_dim, value_dim, self.num_heads).to(device)
        # Cross-attention
        v_prime, _ = cross_attention(ml_c, ml_r, fl_r)
        print(f"After cross_attention - v_prime: {v_prime.shape}")

        # Transformer blocks
        for i, transformer_block in enumerate(self.transformer_blocks):
            v_prime = transformer_block(v_prime)
            print(f"After transformer block {i} - v_prime: {v_prime.shape}")

        # Project to feature dimension
        fl_c = self.output_proj(v_prime)
        print(f"After output projection - fl_c: {fl_c.shape}")

        # Reshape back to 2D
        fl_c = fl_c.transpose(1, 2).view(b, -1, h, w)
        print(f"Final output shape - fl_c: {fl_c.shape}")

        return fl_c

```

**Comment by ueoo** [comment]
> query_dim = ml_c.size(1) # 🤷 this is right?

This is not right.
for a spatial feature `BxCxHxW` should be flattened into `BxCx(HxW)`, and for attention the `(HxW)` is considered as the `L` dim, and `C` is considered as the `D` dim.

**Comment by ueoo** [comment]
Here is a snippet for cross attention, the standard version in Transformer paper, which is implemented based on the https://github.com/lucidrains/vit-pytorch/blob/main/vit_pytorch/simple_vit.py
```python
    def forward(self, queries, keys, values):
        # (b, dim_qk, h, w) -> (b, dim_qk, dim_spatial) -> (b, dim_spatial, dim_qk)
        q = torch.flatten(queries, start_dim=2).transpose(-1, -2)
        q = q + self.q_pos_embedding  # (b, dim_spatial, dim_qk)

        # (b, dim_qk, h, w) -> (b, dim_qk, dim_spatial) -> (b, dim_spatial, dim_qk)
        k = torch.flatten(keys, start_dim=2).transpose(-1, -2)
        k = k + self.k_pos_embedding  # (b, dim_spatial, dim_qk)
        # (b, dim_v, h, w) -> (b, dim_v, dim_spatial) -> (b, dim_spatial, dim_v)
        v = torch.flatten(values, start_dim=2).transpose(-1, -2)

        # (b, dim_spatial, dim_qk) * (b, dim_qk, dim_spatial) -> (b, dim_spatial, dim_spatial)
        dots = torch.matmul(q, k.transpose(-1, -2)) * self.scale

        attn = self.attend(dots)  # (b, dim_spatial, dim_spatial)  softmax

        # (b, dim_spatial, dim_spatial) * (b, dim_spatial, dim_v) -> (b, dim_spatial, dim_v)
        out = torch.matmul(attn, v)
```

For self-attention, it's just a copy of https://github.com/lucidrains/vit-pytorch/blob/main/vit_pytorch/simple_vit.py

**Comment by johndpope** [comment]
@hologerry thanks so much for your pointers - I believe I got something working.
https://wandb.ai/snoozie/IMF/runs/z0m8euqv?nw=nwusersnoozie

if there's an unofficial vgg loss function you could share - or check mine - i would appreciate it. 
im seeing perceptual loss go up.... and image fidelity just konks out.... (though i am overfitting to 1 video - maybe i need to augment the imageset.) 
<img width="1178" alt="Screenshot 2024-07-31 at 7 23 47 pm" src="https://github.com/user-attachments/assets/35b01c3a-ad41-4eba-9247-10f56cc7834e">

**Comment by ueoo** [comment]
@johndpope For the experiment you share, I think it's working. As for perceptual loss I think you should do some hyper parameters tuning, and you can just search `VGGLoss` in Github, most of them are the same, such as https://github.com/CompVis/ipoke/blob/84ceecfefd7bf0769cd35334f2eb824c44c4d56b/utils/losses.py#L67 
You can also try `LPIPS` https://github.com/richzhang/PerceptualSimilarity for a different option.
